### PR TITLE
feat: certs package - part 1

### PR DESCRIPTION
**Pull request by ruseinov**
This introduces general infrastructure for further `certs` package implementation. 
The goal is to be able to validate finality certificates.

### CODE REVIEW

**Review by LesnyRumcajs — DISMISSED** [review]
High-level review:
- unit tests should be implemented. Given it's a completely new code, there's no reason not do it, as we can easily implement our own abstractions to accommodate for them. The coverage should be high. At the current state, this branch make the repository sit at 2% coverage. https://app.codecov.io/github/ChainSafe/rust-f3/tree/ru%2Ffeature%2Fcerts
- the code should be better documented, as outlined in https://www.notion.so/chainsafe/Documentation-practices-1a7ec5e5486b4a56b26d9df6b1677a63

**Comment by ruseinov** [comment]
@LesnyRumcajs unsure how to fix `Potentially unused: strum in gpbft`. It is used for `to_string`. Any way to ignore this?

**Comment by ruseinov** [comment]
Actually, no problem. For now we can just stick with `strum_macros`.

File: gpbft/src/lib.rs
Comment (LesnyRumcajs) [inline gpbft/src/lib.rs:193]: Testing libraries' primitives doesn't bring any value. I'm unsure if it even increases coverage. We should assume the library works as intended.

**Review by LesnyRumcajs — COMMENTED** [review]
Thanks for the tests and docs, it looks much better. I'd skip testing `strum` and anything like this. We should assume that the library works and test only our own logic.

I'll defer to @hanabi1224 for the logic review.

File: gpbft/src/lib.rs
Comment (ruseinov) [reply gpbft/src/lib.rs]: I'm actually unsure if the `UPPER` stuff will ever be critical, but keeping it compatible just in case.
Repr sequence is critical for sure.

**Comment by LesnyRumcajs** [comment]
@hanabi1224

File: gpbft/src/types.rs
Comment (hanabi1224) [inline gpbft/src/types.rs:9]: `PubKey` should derive serde and use base64 encoding

File: gpbft/Cargo.toml
Comment (hanabi1224) [inline gpbft/Cargo.toml:16]:
```suggestion
num-bigint = { version = "0.4.6", features = ["serde"] }
```

File: gpbft/src/types.rs
Comment (hanabi1224) [inline gpbft/src/types.rs:7]: `StoragePower` as part of `PowerEntry` and `PowerTable` should derive serde

File: gpbft/src/lib.rs
Comment (hanabi1224) [inline gpbft/src/lib.rs:83]: `Payload` should derive serde and matches cbor encoding in go-f3
```go
func (t *Payload) MarshalCBOR(w io.Writer) error
```

File: gpbft/src/lib.rs
Comment (hanabi1224) [inline gpbft/src/lib.rs:45]: `SupplementalData` should derive serde and matches cbor encoding in `go-f3`
```go
func (t *SupplementalData) MarshalCBOR(w io.Writer) error
```

File: gpbft/src/api.rs
Comment (hanabi1224) [inline gpbft/src/api.rs:19]: Can we use an enum error type for this trait?

File: gpbft/src/api.rs
Comment (hanabi1224) [inline gpbft/src/api.rs:29]: Ditto, better to use enum error type

File: gpbft/src/api.rs
Comment (hanabi1224) [inline gpbft/src/api.rs:42]: Ditto, on error type

File: gpbft/src/lib.rs
Comment (hanabi1224) [inline gpbft/src/lib.rs:124]: `Justification` should derive serde and match cbor encoding in `go-f3`
```go
func (t *Justification) MarshalCBOR(w io.Writer) error
```

**Comment by ruseinov** [comment]
@hanabi1224 @LesnyRumcajs how about we introduce the ser/de changes in a subsequent PR. I specifically didn't want to tackle that part until it it's used, it's probably easier to deal with once more structs that need it have been introduced. 

This PR mostly tackles the basic logic and some basic abstractions needed for `certs` package implementation. 
That said, `Pubkey` should be a `newtype` in order to implement custom methods, but since it's already a type alias - that would be an easy change once it comes up. 

Good one on the error type, I have not introduced an enum (yet), because there is no need to differentiate between the errors (yet). But I think it's a fair point to introduce it asap to avoid later refactoring.

File: gpbft/src/lib.rs
Comment (hanabi1224) [inline gpbft/src/lib.rs:62]: Phase is encoded as `uint8` in `go-f3` as part of `Payload`

Comment (ruseinov) [reply gpbft/src/lib.rs:62]: Hence the use of `repr(u8)`.

**Comment by ruseinov** [comment]
Since the logic is alright I propose to merge this PR after I have introduce the error enum and continue development in a new one.

**Comment by ruseinov** [comment]
> Since the logic is alright I propose to merge this PR after I have introduce the error enum and continue development in a new one.

Ah I can see now that the comments about error type were made to the interface that's not implemented yet. 
I'll make `Error` an associated type for now. Will deal with this when there is actual logic that needs it.

**Review by hanabi1224 — DISMISSED** [review]

**Review by hanabi1224 — APPROVED** [review]

**Pull request merged**
